src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs

```diff
@@ -77,44 +77,21 @@ public static SourceText From(ITextSnapshot editorSnapshot)
                     throw new ArgumentNullException(nameof(editorSnapshot));
                 }
 
-                return s_textSnapshotMap.GetValue(editorSnapshot, s_createTextCallback);
+                return s_textSnapshotMap.GetValue(editorSnapshot, s => new SnapshotSourceText(s));
             }
 
-            // Use this as a secondary cache to catch ITextSnapshots that have the same ReiteratedVersionNumber as a previously created SnapshotSourceText
-            private static readonly ConditionalWeakTable<ITextBuffer, StrongBox<SnapshotSourceText>> s_textBufferLatestSnapshotMap = new ConditionalWeakTable<ITextBuffer, StrongBox<SnapshotSourceText>>();
-
-            private static SnapshotSourceText CreateText(ITextSnapshot editorSnapshot)
+            /// <summary>
+            /// This only exist to break circular dependency on creating buffer. nobody except extension itself should use it
+            /// </summary>
+            internal static SourceText From(ITextSnapshot editorSnapshot, TextBufferContainer container)
             {
-                var strongBox = s_textBufferLatestSnapshotMap.GetOrCreateValue(editorSnapshot.TextBuffer);
-                var text = strongBox.Value;
-                if (text != null && text._reiteratedVersion == editorSnapshot.Version.ReiteratedVersionNumber)
```

Comment:
this is actual bug where 1:1 mapping got broken.

Reply:
we need to guarantee this. ITextSnapshot(#1) -> SourceText (#2) => ITextSnapshot (#1)

the code above can cause ITextSnapshot(#1) -> SourceText(#2) -> ITextSnapshot (#3)

but if ITextSnapshot (#3) is release, one will one up get null.

Reply:
"the older snapshot"
- it could be older or newer, which ever called AsSourceText first.